spark/src/main/java/com/netflix/iceberg/spark/source/Writer.java

```diff
@@ -202,31 +187,31 @@ public String toString() {
     private final FileFormat format;
     private final String dataLocation;
     private final Map<String, String> properties;
-    private final SerializableConfiguration conf;
     private final String uuid = UUID.randomUUID().toString();
-
-    private transient Path dataPath = null;
+    private final FileIO fileIo;
 
     WriterFactory(PartitionSpec spec, FileFormat format, String dataLocation,
-                  Map<String, String> properties, Configuration conf) {
+                  Map<String, String> properties, FileIO fileIo) {
       this.spec = spec;
       this.format = format;
       this.dataLocation = dataLocation;
       this.properties = properties;
-      this.conf = new SerializableConfiguration(conf);
+      this.fileIo = fileIo;
     }
 
     @Override
     public DataWriter<InternalRow> createDataWriter(int partitionId, long taskId, long epochId) {
       String filename = format.addExtension(String.format("%05d-%d-%s", partitionId, taskId, uuid));
       AppenderFactory<InternalRow> factory = new SparkAppenderFactory();
       if (spec.fields().isEmpty()) {
-        return new UnpartitionedWriter(lazyDataPath(), filename, format, conf.value(), factory);
-
+        return new UnpartitionedWriter(dataLocation, filename, format, factory, fileIo);
       } else {
-        Path baseDataPath = lazyDataPath(); // avoid calling this in the output path function
-        Function<PartitionKey, Path> outputPathFunc = key ->
-            new Path(new Path(baseDataPath, key.toPath()), filename);
+        String baseDataPath = stripTrailingSlash(dataLocation); // avoid calling this in the output path function
+        Function<PartitionKey, String> outputPathFunc = key ->
+            String.format("%s/%s/%s",
+                stripTrailingSlash(baseDataPath),
+                stripTrailingSlash(stripLeadingSlash(key.toPath())),
```

Comment:
This is always implemented by [`PartitionSpec#partitionToPath`](https://github.com/apache/incubator-iceberg/blob/master/api/src/main/java/com/netflix/iceberg/PartitionSpec.java#L143-L156). That never adds a starting or trailing `/` so this can rely on that and doesn't need to do the extra check.